docs/contracts/v4/guides/08-custom-accounting.mdx

```diff
@@ -0,0 +1,235 @@
+
+```solidity
+function beforeSwap(
+    address,
+    PoolKey calldata key,
+    IPoolManager.SwapParams calldata params,
+    bytes calldata
+) external override returns (bytes4, BeforeSwapDelta, uint24) {
+		// Implementation details will be explained in the following sub-steps
+}
+
+```
+
+### Step 2.1: Calculate the swap amount and fee
+
+We determine the absolute swap amount and calculate our fee based on it.
+
+```solidity
+uint256 swapAmount = params.amountSpecified < 0
+    ? uint256(-params.amountSpecified)
+    : uint256(params.amountSpecified);
+uint256 feeAmount = (swapAmount * HOOK_FEE_PERCENTAGE) / FEE_DENOMINATOR;
+
+```
+
+### Step 2.2: Collect the fee
+
+We use `poolManager.take` to collect the fee. This creates a debt for our hook in the specified currency.
+
+```solidity
+Currency feeCurrency = params.zeroForOne ? key.currency0 : key.currency1;
+poolManager.take(feeCurrency, address(this), feeAmount);
+
+```
+
+### Step 2.3: Create the BeforeSwapDelta
+
+This is where the magic happens. We create a `BeforeSwapDelta` that represents our fee. The `toBeforeSwapDelta` function takes two parameters:
+
+- The specified delta: This is our fee amount. It's positive because we're adding it to the hook's balance.
+- The unspecified delta: We set this to 0 as we're not affecting the other currency.
+
+```solidity
+BeforeSwapDelta returnDelta = toBeforeSwapDelta(
+    int128(int256(feeAmount)),// Specified delta (fee amount)
+    0// Unspecified delta (no change)
+);
+
+```
+
+### Step 2.4: Return values
+
+We return the function selector, our `returnDelta`, and 0 for the fee override.
+
+```solidity
+return (BaseHook.beforeSwap.selector, returnDelta, 0);
+```
+
+## Step 3: Understanding the BeforeSwapDelta Mechanism
+
+Now, let's dive deeper into how the `BeforeSwapDelta` works and how it affects the overall swap process.
+
+1. **Initial State**:
+Let's say a user wants to swap 100 USDC for USDT. Initially:
+    - Hook's delta: (0, 0)
+    - User's swap request: -100 USDC (negative because they're selling)
+    
+2. **After Hook Execution**:
+Our hook has taken a 1 USDC fee (assuming 1% for simplicity):
+    - Hook's delta: (-1 USDC, 0) // The hook now owes 1 USDC to the pool
+    - BeforeSwapDelta returned: (1 USDC, 0) // This will be added to the hook's delta and subtracted from the swap delta
+    
+3. **PoolManager Processing**:
+The PoolManager applies our `BeforeSwapDelta`:
+The pool then swaps 99 USDC for, let's say, 98 USDT.
+    
+    ```solidity
+    amountToSwap = params.amountSpecified + hookDelta.getSpecifiedDelta();
+    -99 USDC = -100 USDC + 1 USDC
+    ```
+    
+4. **Delta Resolution**:
+The PoolManager then resolves the deltas:
+    
+    ```solidity
+    // Hook's new delta
+    newHookDelta = oldHookDelta + returnDelta
+    (0, 0) = (-1 USDC, 0) + (1 USDC, 0)
+    
+    // Swap delta for router
+    swapDelta = (-99 USDC, 98 USDT) - (1 USDC, 0)
+              = (-100 USDC, 98 USDT)
+    ```
```

Comment:
```suggestion
```solidity
// Hook's new delta
newHookDelta = oldHookDelta + returnDelta
(0, 0) = (-1 USDC, 0) + (1 USDC, 0)
// Swap delta for router
swapDelta = (-99 USDC, 98 USDT) - (1 USDC, 0)
          = (-100 USDC, 98 USDT)
```

```

Reply:
can you double check this fix locally? the suggestion doesnt look right to me since its missing the closing ticks ```

github doesnt seem to like it haha